pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,361 @@
+package tfbridge
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+type PulumiInputValidator struct {
+	// The resource URN that we are validating
+	urn resource.URN
+
+	// The pulumi schema of the package
+	schema pschema.PackageSpec
+}
+
+type TypeFailure struct {
+	// The Reason for the type failure
+	Reason string
+
+	// The path to the property that failed
+	ResourcePath string
+}
+
+// NewInputValidator creates a new input validator for a given resource and
+// package schema
+func NewInputValidator(urn resource.URN, schema pschema.PackageSpec) *PulumiInputValidator {
+	return &PulumiInputValidator{
+		urn:    urn,
+		schema: schema,
+	}
+}
+
+func propertiesWithDefaults(
+	propertyTypes map[string]pschema.PropertySpec,
+) []resource.PropertyKey {
+	keysWithDefaults := []resource.PropertyKey{}
+	for key, spec := range propertyTypes {
+		if spec.Default != nil {
+			keysWithDefaults = append(keysWithDefaults, resource.PropertyKey(key))
+
+		}
+	}
+	return keysWithDefaults
+}
+
+// validatePropertyValue is the main function for validating a PropertyMap against the Pulumi Schema. It is a
+// recursive function that will validate nested types and arrays. Returns a list of type failures if any are found.
+func (v *PulumiInputValidator) validatePropertyMap(
+	propertyMap resource.PropertyMap,
+	required []string,
+	propertyTypes map[string]pschema.PropertySpec,
+	propertyPath resource.PropertyPath,
+) []TypeFailure {
+	stableKeys := propertyMap.StableKeys()
+	failures := []TypeFailure{}
+
+	// handle required properties
```

Comment:
While this looks right, I'm still worried here we actually allow some cases where "required" properties are actually filled out by the bridge post-check and therefore this will be too strict. I can try constructing some examples.

Reply:
Yeah I'd be more comfortable starting with that.

Reply:
Done!